lib/stateView/document/query/validateOrderByFields.js

```diff
@@ -0,0 +1,130 @@
+ * @param {Array} [whereConditions]
+ * @returns {Array}
+ */
+function validateOrderByFields(
+  dataContractIndexFields,
+  orderByConditions,
+  whereConditions = [],
+) {
+  // flat where conditions and set sort directions
+  const compiledWhereBy = whereConditions.reduce(
+    (fields, [field, operator, elementMatchValue]) => {
+      let fieldsToAdd = [];
+
+      if (operator === 'elementMatch') {
+        fieldsToAdd = elementMatchValue.map(([item]) => [
+          `${field}.${item}`,
+          '*',
+        ]);
+      } else {
+        fieldsToAdd = [[field, '*']];
+      }
+
+      return fields.concat(fieldsToAdd);
+    },
+    [],
+  );
+
+  // convert dataContractIndex index objects to array
+  let compiledDataContractIndexFields = dataContractIndexFields.map(
+    compoundIndex => compoundIndex.map((index) => {
+      const [[field, direction]] = Object.entries(index);
+
+      return [field, direction];
+    }),
+  );
+
+  // mongoDB can sort by two directions in single field indices. So we can add this directions.
+  const additionalIndices = [];
+  compiledDataContractIndexFields.forEach((compoundIndex) => {
+    if (compoundIndex.length === 1) {
+      const [field, direction] = compoundIndex[0];
+      const newDirection = direction === 'asc' ? 'desc' : 'asc';
+      const alreadyHasIndex = compiledDataContractIndexFields.some(
+        (indexArray) => {
+          const [oldField, oldDirection] = indexArray[0];
+
+          return (
+            indexArray.length === 1
+            && oldField === field
+            && oldDirection === newDirection
+          );
+        },
+      );
+
+      if (!alreadyHasIndex) {
+        additionalIndices.push([[field, newDirection]]);
+      }
+    }
+  });
+
+  compiledDataContractIndexFields = compiledDataContractIndexFields.concat(
+    additionalIndices,
+  );
+
+  // leave only indices that contain all orderBy fields
+  compiledDataContractIndexFields = compiledDataContractIndexFields.filter(
+    compoundIndex => orderByConditions
+      .every(([field, direction]) => compoundIndex
+        .some(
+          ([indexField, indexDirection]) => field === indexField && direction === indexDirection,
+        )),
+  );
+
+  // leave only only fields that not pass out validation
+  return orderByConditions
+    .filter(([field]) => (
+      !compiledDataContractIndexFields
+        .find(index => index
+          .find(([indexField], i) => {
+            if (indexField !== field) {
+              return false;
+            }
+
+            // part of compound index before our orderBy field
+            const compoundIndexBefore = index.slice(0, i);
+
+            // part of compound index after our orderBy field
+            const compoundIndexAfter = index.slice(i, index.length);
+
+            // check that all index fields of this part are inside where or orderBy conditions
+            const compoundFieldsBeforeIsOk = checkIndices(compoundIndexBefore, compiledWhereBy)
```

Comment:
Let's have a separate function for `areIndexedFieldsPresentInWhereCondition(compoundIndexBefore, compiledWhereBy)`